### PR TITLE
Allow params once again in embeds

### PR DESCRIPTION
**Pull request by ziahamza**
An attempt to re-enable params in embeds.

### CODE REVIEW

**Comment by stevben** [comment]
+1 !
I can be a beta tester if you want. I have a great use case for that.

**Comment by ziahamza** [comment]
@stevben That would be great. We have also deployed this internally, and is working without bugs for the most part :)

**Comment by arikfr** [comment]
"for the most part" -- there is a part it has bugs with? :)

**Comment by ziahamza** [comment]
@arikfr The page shows an `Internal failure` when the params are not passed, as it crashes it. Should probably show a more appropriate error message as its not obvious. 

I can add a fix for ^ if the rest of the PR looks good

File: client/app/pages/queries/visualization-embed.js [hunk 1]
```diff
@@ -33,7 +33,8 @@ export default function (ngModule) {
     return session($http, $route, Auth).then(() => {
       const queryId = $route.current.params.queryId;
       const query = $http.get(`/api/queries/${queryId}`).then(response => response.data);
-      const queryResult = $http.get(`/api/queries/${queryId}/results.json`).then(response => response.data);
+      // pass down any params from the top level url
+      const queryResult = $http.post(`/api/queries/${queryId}/results.json${location.search}`).then(response => response.data);
```
Comment (arikfr) [inline client/app/pages/queries/visualization-embed.js]: Let's keep the old behavior and only use the new one, if it's enabled via [the feature flag](https://github.com/getredash/redash/blob/master/redash/settings.py#L236).

File: redash/handlers/query_results.py [hunk 2]
```diff
@@ -101,6 +151,28 @@ def options(self, query_id=None, query_result_id=None, filetype='json'):
         return make_response("", 200, headers)
 
     @require_permission('view_query')
+    def post(self, query_id=None, filetype='json'):
```
Comment (arikfr) [inline redash/handlers/query_results.py]: This should return error 400 if the feature flag is disabled, as it's a security concern. Worth adding a comment here about this as well.

File: client/app/pages/queries/visualization-embed.js (hunk 1 above)
Comment (ziahamza) [inline client/app/pages/queries/visualization-embed.js]: @arikfr Question: whats the best way to access featureFlags from the client side? Or should I pass in a param to the backend, and check the featureFlag from there?

Comment (ziahamza) [inline client/app/pages/queries/visualization-embed.js]: resolve it by checking it on the server side 👍

**Comment by ziahamza** [comment]
@arikfr updated the PR to only run queries when the featureFlag is set. Should be good to go now 👍

**Comment by ziahamza** [comment]
@arikfr any updates on getting this PR upstream?

**Comment by rohanpd** [comment]
We've been beta testing this and have come across some embeds that appear to time out still (regardless on whether they have parameters or not):

Failed to load resource: the server responded with a status of 502 (Bad Gateway)

**Comment by ziahamza** [comment]
@rohanpd There is a timeout somewhere I think. For that I think it will break either way (either by redash or the http reverse proxies in general in front of redash).

**Comment by ziahamza** [comment]
@arikfr would love to get your comments on this and push this to upstream.

**Comment by morganizeit** [comment]
What's the status of this pull request? No updates for a month.

My org has ~160 parameterized embeds to facilitate operations. This is for fast reference of account details, eliminating multiple clicks.

Clicks become seconds become minutes... Since you can't put {{params}} into embedded views the GET params are the only way to filter a view without sending people out to the ReDash site, losing the context of where they were at. Embedding is important for getting the most from a reporting platform like ReDash.

It worries me that big parts of the embedding API can go missing without warning. For now our workaround is to manually edit all the iframes to point at an older version of ReDash in Staging. Lots of work and must be repeated when the bugfix is ready.

**Comment by arikfr** [comment]
Now that we have a stable 1.0 release, we can get back to working on getting this into master.

@ziahamza:

> @rohanpd There is a timeout somewhere I think. For that I think it will break either way (either by redash or the http reverse proxies in general in front of redash).

There is a timeout in Gunicorn and might be in other proxies upstream (like ELB/ALB). But the problem introduced here is that previously queries without parameters were using the cached version of the query, hence not hitting the timeout. This pull request changed this behavior for all queries, instead of changing it for only queries with parameters.

Basically this pull request should work like the previous solution: introduce a change in behavior only when the feature flag is toggled and only to parameterized queries.

**Comment by arikfr** [comment]
@morganizeit I understand your frustration, but just want to clarify something:

> It worries me that big parts of the embedding API can go missing without warning. 

1. This part of the API was always considered non production ready and was behind a feature flag for this reason.
2. "without warning" - it was made clear in the release notes, that the 1.0 release doesn't have support for this feature and that it will be introduced back after the 1.0 release.

**Comment by morganizeit** [comment]
Hey thanks Arik! Just back from travel and catching up. Re-reading my comment it sounds pretty nasty. Yuck, please accept my apologies.

I'm not an admin of the system so I did not know this was a beta feature, off by default. I came across this thread and the release notes but did not understand that it was not _previously_ supported either, I thought it simply changed to unsupported in this release when it was dropped: https://discuss.redash.io/t/embed-visualization-with-parameters/477

Thanks for the clarification and for warming up the feature again. Parameterized embed enables showing data where people are already looking, increasing awareness of ReDash and what it can do.

Cheers,
Morgan.

**Comment by nt3478** [comment]
+1

**Comment by morganizeit** [comment]
Hi folks! Any updates RE sorting out obstacles for this feature?

**Comment by ziahamza** [comment]
@arikfr updated the PR to only run queries from cache if query parameters are not passed. https://github.com/getredash/redash/pull/1524/files#diff-f734d32676ec526e17e5ae92d41b5787R193

does this work?

**Comment by problem** [comment]
Another +1 on this!

**Comment by arikfr** [comment]
Thanks @ziahamza , I will review it again.

**Comment by morganizeit** [comment]
Aaaaalmost there... all the CodeClimate errors are about whitespace and line length. @ziahamza added the feature flag. Are there any other obstacles?

**Comment by azpeters** [comment]
greatly anticipating this. any estimate for its planned release date?

**Comment by morganizeit** [comment]
Our data team applied the PR changes to our internal fork and deployed, looking good so far. Thank you for working to bring this beta feature back!

Cheers,
Morgan.

File: redash/handlers/query_results.py [hunk 3]
```diff
@@ -9,14 +10,63 @@
...
 
 
+#
+# Run a parameterized query synchronously and return the result
+# DISCLAIMER: Temporary solution to support parameters in queries. Should be
+#             removed once we refactor the query results API endpoints and handling
+#             on the client side. Please don't reuse in other API handlers.
+#
+def run_query_sync(data_source, parameter_values, query_text, max_age=0):
+    query_parameters = set(collect_query_parameters(query_text))
```
Comment (ziahamza) [inline redash/handlers/query_results.py]: its the same as for the usual `run_query` as well: https://github.com/getredash/redash/pull/1524/files#diff-f734d32676ec526e17e5ae92d41b5787R71

**Comment by phpcodecrafting** [comment]
Can we please prioritize this? It's been open for 5+ months. Additionally the last 3 releases have all incorrectly stated that the "next" release would bring `ALLOW_PARAMETERS_IN_EMBEDS` back.

**Comment by arikfr** [comment]
@phpcodecrafting I meant the next major release (v2).

Anyway, merged.

**Comment by arikfr** [comment]
@ziahamza thank you for your help here 👍

**Comment by phpcodecrafting** [comment]
Awesome!  Many thanks to all involved!
